### ISSUE TITLE
adapters produce unk tokens only

### ISSUE THREAD

**Issue by flozi00** [predibase/lorax#68]
### System Info

docker from latest

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [ ] An officially supported command
- [ ] My own modifications

### Reproduction

example prompt:
```

User: Who are you ? 

Assistant: You mean me ?

User: Who are you ? 

Assistant: You mean me ?

User: Who are you ? 

Assistant: You mean me ?

User: Who are you ? 

Assistant: You mean me ?

User: Who are you ? 

Assistant: You mean me ?
```

this is enough to produce unk tokens on any prompt only.
When cutting down to 2 lines only it works as expected.

base model is:
mistralai/Mistral-7B-v0.1

adapters tested:
https://huggingface.co/qblocks/mistral_7b_norobots
https://huggingface.co/flozi00/mistral-zephyr-lora
https://huggingface.co/flozi00/mistral-germanassistantv4

start command

```
docker run --pull always --gpus all -d --shm-size 1g -p 8080:80 ghcr.io/predibase/lorax:latest --model-id mistralai/Mistral-7B-Instruct-v0.1 --cuda-memory-fraction 0.5 --max-total-tokens 8192 --max-batch-prefill-tokens 7000 --max-input-length 7000
```

### Expected behavior

Working with and without adapters loaded

**Comment by flozi00** [comment]
@tgaddair could you please take a look when you have time for ?

Already checked several ideas, but no progress

**Comment by flozi00** [comment]
Also checked 0.1.2 release, container, same there

**Comment by runvnc** [comment]
Is it possible this has something to do with the 4bit quantization not being compatible somehow? Or maybe lorax doesn't really support all of the target modules? I also had trouble with a non-quantized fine-tune which seemed to produce no output. Although, I might have done something wrong, since I was trying to modify the qlora code to take the quantization out. But mine also had extra target modules besides the four listed in the README, which I suspect might be an issue.

BTW, @flozi00, thanks for your work on the 4bit. Am I understanding correctly that this actually is working for you, just not with certain prompts?

**Comment by tgaddair** [comment]
@flozi00 here's the PR to allow disabling SGMV temporarily: #69.

**Comment by flozi00** [comment]
@runvnc the error comes without qunatization, the 4bit PR is independent because i told to take this feature

@tgaddair 
1. thanks a lot
2. it was working when i did my first tests with lorax, the unsupported modules just got dropped when i understand correctly, so the quality had small loss but still was working. It could be working with solving 1. but just not as good as expected :)

**Comment by abhijithnair1** [comment]
@tgaddair By disabling the `SGMV`, how much of a performance drop can we expect?

**Comment by flozi00** [comment]
![image](https://github.com/predibase/lorax/assets/47894090/e24c694d-63a1-40d8-8b4f-3127321e0a6b)
![image](https://github.com/predibase/lorax/assets/47894090/21be4e0c-33c6-4523-97f8-0173c9fe8ccc)

Still having the unk error, with disabled sgmv kernel

**Comment by tgaddair** [comment]
@abhijithnair1, the performance drop should be very minimal below 8 concurrent adapters. It only becomes noticeable with a high number of adapters at once. That said, I hope to have a solution in place by end of day to make sure SGMV is working for rank 32 adapters.

@flozi00 I wonder if it's related to the prompt at all. The prompt I was using was this one, can you try and verify it generates something sensible?

```
curl 127.0.0.1:8080/generate \
    -X POST \
    -d '{"inputs": "<|system|> You are a helpful assistant <|user|> What is deep learning? </s> <|assistant|>", "parameters": {"max_new_tokens": 64, "adapter_id": "qblocks/mistral_7b_norobots"}}' \
    -H 'Content-Type: application/json'
```

**Comment by flozi00** [comment]
Now its just empty output, but no unk tokens
Maybe something related to my dev env

**Comment by tgaddair** [comment]
@flozi00 @abhijithnair1 I believe I have a fix for the issue with the rank 32 adapters in #70 if you want to test it out. Updating to a newer version of the SGMV shrink kernel appears to have solved the issue on my end.

**Comment by runvnc** [comment]
I also did a test the other day and had my adapter load successfully but
did not see any output. This was on a prebuilt image mentioned in the
README, not a newly built image.

**Comment by tgaddair** [comment]
Hey @runvnc can you share your adapter and example prompt here to help with reproducing the issue?

I've tested #70 with a variety of different adapters and configs and feel confident about merging it once the PR updates with my latest changes (I think GH is experiencing some degraded performance at the moment). So hopefully we'll have a fix for the rank 32 adapters shortly.

**Comment by runvnc** [comment]
I will have to test again today with the new docker version. This is what I am trying to ask about: 
 
 ``` python
    # save tokenizer for easy inference
    tokenizer = AutoTokenizer.from_pretrained(args.model_id)
    tokenizer.save_pretrained(sagemaker_save_dir)```
    
    This creates files like `special_tokens_map.json`. Are special tokens supported?

**Comment by tgaddair** [comment]
Latest image with GMV fix is up if foljs want to test it out: https://github.com/orgs/predibase/packages/container/lorax/151913146?tag=f9409af

**Comment by tgaddair** [comment]
@runvnc that's an interesting one. Are these special tokens that were added during fine-tuning of the adapter? If so, then it's quite likely we aren't properly accounting for those. We would likely need to update the decoding step to handle this. Is there a publicly available adapter in HF you could share that I can use to repro? Happy to add a separate issue to track this.

**Comment by runvnc** [comment]
@tgaddair https://huggingface.co/qblocks/mistral_7b_norobots/tree/main 

This one, was mentioned at the top of the issue. The main qLoRA repo example uses this AutoTokenizer thing.

**Comment by tgaddair** [comment]
Thanks @runvnc, I'll file an issue to track fixing this.

**Comment by runvnc** [comment]
Seems like it just might be related though, since this issue mentions tokens.

**Comment by tgaddair** [comment]
@runvnc yes I think you're right for at least that model. For the prompt I was using, the issue was due to the SGMV kernel, but there could certainly be other issues that stem from the tokenizer.

**Comment by runvnc** [comment]
@flozi00 if you have code like this https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L346C9-L346C9 using the AutoTokenizer, you might try removing that, and making sure there is nothing about special_tokens in model output dir. Since those special tokens seem to not be supported yet. I think that might have been one of my problems. Right now my adapter seems to work (at least generate text) after I took that out and removed quantization stuff from the qlora training code.

**Comment by flozi00** [comment]
I would close this issue, since the latest commit all adapters are working fine
